nighthawk/source/common/statistic_impl.cc

```diff
@@ -0,0 +1,216 @@
+      a.sum_of_squares_ + b.sum_of_squares_ +
+      pow(a.mean() - b.mean(), 2) * a.count() * b.count() / combined->count();
+  return combined;
+}
+
+InMemoryStatistic::InMemoryStatistic() : streaming_stats_(std::make_unique<StreamingStatistic>()) {}
+
+void InMemoryStatistic::addValue(int64_t sample_value) {
+  samples_.push_back(sample_value);
+  streaming_stats_->addValue(sample_value);
+}
+
+uint64_t InMemoryStatistic::count() const {
+  ASSERT(streaming_stats_->count() == samples_.size());
+  return streaming_stats_->count();
+}
+double InMemoryStatistic::mean() const { return streaming_stats_->mean(); }
+double InMemoryStatistic::variance() const { return streaming_stats_->variance(); }
+double InMemoryStatistic::stdev() const { return streaming_stats_->stdev(); }
+
+std::unique_ptr<Statistic> InMemoryStatistic::combine(const Statistic& statistic) {
+  auto combined = std::make_unique<InMemoryStatistic>();
+  const InMemoryStatistic& b = dynamic_cast<const InMemoryStatistic&>(statistic);
+
+  combined->samples_.insert(combined->samples_.end(), this->samples_.begin(), this->samples_.end());
+  combined->samples_.insert(combined->samples_.end(), b.samples_.begin(), b.samples_.end());
+  combined->streaming_stats_ = this->streaming_stats_->combine(*b.streaming_stats_);
+  return combined;
+}
+
+const int HdrStatistic::SIGNIFICANT_DIGITS = 4;
+
+HdrStatistic::HdrStatistic() : histogram_(nullptr) {
+  // Upper bound of 60 seconds (tracking in nanoseconds).
+  const uint64_t max_latency = 1000L * 1000 * 1000 * 60;
+
+  int status = hdr_init(1 /* min trackable value */, max_latency, HdrStatistic::SIGNIFICANT_DIGITS,
+                        &histogram_);
+  if (status != 0) {
+    ENVOY_LOG(error, "Failed to initialize HdrHistogram.");
+    histogram_ = nullptr;
+  }
+}
+
+// TODO(oschaaf): valgrind complains when a Histogram is created but never used.
+HdrStatistic::~HdrStatistic() {
+  if (histogram_ != nullptr) {
+    hdr_close(histogram_);
+    histogram_ = nullptr;
+  }
+}
+
+void HdrStatistic::addValue(int64_t value) {
+  if (histogram_ != nullptr) {
+    if (!hdr_record_value(histogram_, value)) {
+      ENVOY_LOG(warn, "Failed to record value into HdrHistogram.");
+    }
+  }
+}
+
+uint64_t HdrStatistic::count() const { return histogram_->total_count; }
+double HdrStatistic::mean() const { return hdr_mean(histogram_); }
+double HdrStatistic::variance() const {
+  return stdev() * stdev();
+  ;
+}
+double HdrStatistic::stdev() const {
+  // HdrHistogram_c's stdev actually gives us the population standard deviation.
+  // So we compute the sample standard deviation ourselves instead.
+  // TODO(oschaaf): this fixes some of the test expectations, but figure out if
+  // stdev or pstdev is preferrable. Looks like wrk2 uses pstdev which would produce
+  // (slightly) better numbers, though that probably isn't a reason for us to decice
+  // which one to use here. Switching to pstdev would get rid of having to do this
+  // ourselves.
```

Comment:
When measuring latency of a real system, i.e. just probing, arguably we are just sampling. OTOH, when doing load tests, all the traffic is from the load generator, so could be thought of as the population. Not sure it's hugely important though as long as we document it, we're just after some idea of variance. If we can simplify the below, it sounds like a win.